### PR TITLE
DM-54487: Fix overflow in afwMath binImage

### PR DESCRIPTION
**Pull request by taranu**

### CODE REVIEW

File: src/math/binImage.cc [hunk 1]
```diff
@@ -47,36 +47,65 @@ std::shared_ptr<ImageT> binImage(ImageT const& in, int const binX, int const bin
...
     std::shared_ptr<ImageT> out =
             std::shared_ptr<ImageT>(new ImageT(lsst::geom::Extent2I(outWidth, outHeight)));
     out->setXY0(in.getXY0());
     *out = typename ImageT::SinglePixel(0);
 
-    for (int oy = 0, iy = 0; oy < out->getHeight(); ++oy) {
-        for (int i = 0; i != binY; ++i, ++iy) {
+    // Store remainders from integer binning so that accumulation can be done
+    // per-row, limiting the opportunity for integer overflow
+    std::unique_ptr<std::vector<int>> remainders;
```
Comment (mwittgen) [inline src/math/binImage.cc]: Too much indirection.

```suggestion
    std::vector<int> remainders;
    if constexpr (is_integer) {
        remainders.resize(outWidth);
    }
```

File: src/math/binImage.cc [hunk 2]
```diff
@@ -47,36 +47,65 @@ std::shared_ptr<ImageT> binImage(ImageT const& in, int const binX, int const bin
...
             typename ImageT::x_iterator optr = out->row_begin(oy);
-            for (typename ImageT::x_iterator iptr = in.row_begin(iy), iend = iptr + binX * outWidth;
+            for (typename ImageT::x_iterator iptr = in.row_begin(iy), iend = iptr + binX_u * outWidth;
                  iptr < iend;) {
-                typename ImageT::SinglePixel val = *iptr;
-                ++iptr;
-                for (int j = 1; j != binX; ++j, ++iptr) {
+                typename ImageT::SinglePixel val = *(iptr++);
+                for (unsigned int j = 1; j != binX_u; ++j, ++iptr) {
                     val += *iptr;
```
Comment (mwittgen) [inline src/math/binImage.cc]: Can `val` overflow for `uint16_t` ?

File: src/math/binImage.cc [hunk 3]
```diff
@@ -47,36 +47,65 @@ std::shared_ptr<ImageT> binImage(ImageT const& in, int const binX, int const bin
...
 
-    int const outWidth = in.getWidth() / binX;
-    int const outHeight = in.getHeight() / binY;
+    unsigned int binX_u = binX;
+    unsigned int binY_u = binY;
+
+    unsigned int const outWidth = in.getWidth() / binX_u;
+    unsigned int const outHeight = in.getHeight() / binY_u;
+
+    const typename ImageT::SinglePixel binXY = binX_u * binY_u;
```
Comment (mwittgen) [inline src/math/binImage.cc]: For uint16_t the bin product can overflow

**Review by mwittgen — COMMENTED** [review]
The existing version already relies on implicit integer promotion. I am not sure of fir the `uint16_t` case no overflows occur. Some unit tests for the corner cases could be added.

**Review by mwittgen — APPROVED** [review]
